### ISSUE TITLE
Error "Variable "$id" is never used" not fail build process

### ISSUE THREAD

**Issue by Mangese** [apollographql/apollo-ios#1344]
when generate API.swift with graphql file that has unuse variables.
an error `Variable "xxx" is never used in operation "xxx"` is log but build process is success and run an app normally
```
query Repo($id: ID!) {
    currentuser {
        id
    }
}

```

but on other error, the build process will fail 

Ex. generate API.swift with below graphql result an error 

`error: Cannot query field "something" on type "Query". Did you mean "setting"?` 

and build process is fail
```
query Repo {
    currentuser {
        id
    }
    something
}
```

PS: Apollo version 0.27.1

**Comment by designatednerd** [comment]
I would first recommend updating to a more recent version of the SDK - IIRC this was a bug in the underlying CLI that has since been addressed. If that doesn't help, let me know.

**Comment by Mangese** [comment]
update to Apollo 0.30.0 doesn't help.

**Comment by Mangese** [comment]
```
Variable "$id" is never used in operation "Repo".

Command PhaseScriptExecution emitted errors but did not return a nonzero exit code to indicate failure
```


Error from log when build with

```
query Repo($id: ID!) {
    currentuser {
        id
    }
}
```

**Comment by designatednerd** [comment]
OK - are you using the shell script or the Swift Scripting method to run your codegen?

**Comment by Mangese** [comment]
I'm not sure about that.

I add a run script phase on Build Phase that has script below with shell = sh 
```
PATH=$PATH:/usr/local/opt/node@8/bin
SCRIPT_PATH="${PODS_ROOT}/Apollo/scripts"

if [ -z "${XCS_BOT_NAME}" ]; then

cd "${SRCROOT}/${TARGET_NAME}"
"${SCRIPT_PATH}"/run-bundled-codegen.sh codegen:generate --target=swift --includes=./GraphQL/**/*.graphql --localSchemaFile="GraphQL/Generated/schema.json" GraphQL/Generated/API.swift --customScalarsPrefix=GraphQL
fi

```

**Comment by designatednerd** [comment]
Yep that's the shell script method. Can you confirm that in your schema `currentuser` requires a non-optional `id`, meaning an `ID!`? If it only requires an `ID` (without the exclamation point) then that might be the source of the issue.

**Comment by Mangese** [comment]
`currentuser ` does not require any params. an `$id: ID!` is just a unuse parameter that i forgot to remove. it can change to any params Ex. ` Repo($abcd: String!)` or ` Repo($test: Int!)`

if RepoQuery never called or called with given parameter Ex. `RepoQuery(abcd: "a")` it'll build success with error that mention before but will throw an  `Variable $abcd is never used in operation "Repo" ` from the server

**Comment by designatednerd** [comment]
> `currentuser` does not require any variable

I'm almost positive this is why it's not exiting, then. When the error is more of a "Hey, you forgot to remove this" rather than a "Nope, this is not a valid query", we don't want to fail the entire build.

**Comment by Mangese** [comment]
I understand that it's not a fatal error for an entire app. but if the query is sent to graphql server it'll always fail due to graphql default validation rule(if I understand it correctly). 
So, I think it's not different from invalid query because the query always fails if sends to server

**Comment by designatednerd** [comment]
Hm...what server are you using? I think that might be a server implementation detail, but I could be wrong.

**Comment by Mangese** [comment]
`apollo-server-express` from https://github.com/apollographql/apollo-server/tree/main/packages/apollo-server-express

**Comment by Mangese** [comment]
`GraphQL operation is only valid if all variables defined by an operation are used `
from graphql-js [line](https://github.com/graphql/graphql-js/blob/80471134f1152c69da2cc27518a127991cc801b9/src/validation/rules/NoUnusedVariablesRule.d.ts#L7
)

**Comment by designatednerd** [comment]
Ahhh got it. OK, will talk to some tooling folks on this.

**Comment by Mangese** [comment]
thanks!

**Comment by designatednerd** [comment]
(FYI that team is super backed up, so it may take some time)

**Comment by designatednerd** [comment]
So our tooling folks are working on a new Rust-based parsing and validation of schema + queries, it will have to land with that. 

I believe the reason it's not in there right now is because it would cause errors when people were using it in watch mode (mostly useful for JS) and it got disabled for that. However, I'm going to be working to make sure there's at least a mode that can be selected in new Rust stuff to bomb out on errors like this.

**Comment by calvincestari** [comment]
Hi 👋🏻 - this issue should be resolved in 1.0. 1.0 is currently in alpha releases with a full 1.0 release in H2 2022. Please take a look at the [latest alpha](https://github.com/apollographql/apollo-ios/releases) and let us know if this is still an issue for you.
